azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/IllegalEventDataBodyException.java

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) Microsoft. All rights reserved.
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+package com.microsoft.azure.eventhubs;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.qpid.proton.amqp.messaging.AmqpSequence;
+import org.apache.qpid.proton.amqp.messaging.AmqpValue;
+import org.apache.qpid.proton.amqp.messaging.Data;
+
+import com.microsoft.azure.servicebus.amqp.AmqpConstants;
+
+public class IllegalEventDataBodyException extends RuntimeException {
+    
+    private static final Map<Class, String> KNOWN_SECTIONS = new HashMap<Class, String>() {{
+        put(AmqpValue.class, AmqpConstants.AMQP_VALUE);
+        put(AmqpSequence.class, AmqpConstants.AMQP_SEQUENCE);
+    }};
+    
+    private final Class bodySection;
+    
+    public IllegalEventDataBodyException(final Class actualBodySection) {
+        super(KNOWN_SECTIONS.containsKey(actualBodySection)
+            ? String.format("AmqpMessage Body Section will be available in %s.getBody() only if it is of type: %s. If AmqpMessage is sent with any other Body type - it will be added to %s.getSystemProperties(). Use thisException.getSystemPropertyName() method to find this value in %s.getSystemProperties()",
+                    EventData.class, Data.class, EventData.class, EventData.class)
+            : "AmqpMessage Body Section cannot be mapped to any EventData section.");
+        this.bodySection = actualBodySection;
```

Comment:
While as per the SPEC this is correct - this happened a while back: -we saw ApplicationProperties as part of Data section (I am still investigating how this was possible):

17/02/02 20:34:20 ERROR ReceiverTracker: Deregistered receiver for stream 19: Restarting receiver with delay 2000ms: Error handling message, restarting receiver for partition 19 - java.util.concurrent.ExecutionException: java.lang.ClassCastException: org.apache.qpid.proton.amqp.messaging.ApplicationProperties cannot be cast to org.apache.qpid.proton.amqp.messaging.Data
        at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at org.apache.spark.streaming.eventhubs.EventHubsClientWrapper.receive(EventHubsClientWrapper.scala:158)
        at org.apache.spark.streaming.eventhubs.EventHubsReceiver$EventHubsMessageHandler.run(EventHubsReceiver.scala:117)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassCastException: org.apache.qpid.proton.amqp.messaging.ApplicationProperties cannot be cast to org.apache.qpid.proton.amqp.messaging.Data
        at com.microsoft.azure.eventhubs.EventData.<init>(EventData.java:85)
        at com.microsoft.azure.eventhubs.EventDataUtil.toEventDataCollection(EventDataUtil.java:46)
        at com.microsoft.azure.eventhubs.PartitionReceiver$3.apply(PartitionReceiver.java:285)
        at com.microsoft.azure.eventhubs.PartitionReceiver$3.apply(PartitionReceiver.java:281)
        at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:602)
        at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
        at com.microsoft.azure.servicebus.MessageReceiver$4.onEvent(MessageReceiver.java:305)
        at com.microsoft.azure.servicebus.amqp.DispatchHandler.onTimerTask(DispatchHandler.java:10)
        at com.microsoft.azure.servicebus.amqp.ReactorDispatcher$ScheduleHandler.run(ReactorDispatcher.java:126)
        at org.apache.qpid.proton.reactor.impl.SelectableImpl.readable(SelectableImpl.java:118)
        at org.apache.qpid.proton.reactor.impl.IOHandler.handleQuiesced(IOHandler.java:61)
        at org.apache.qpid.proton.reactor.impl.IOHandler.onUnhandled(IOHandler.java:381)
        at org.apache.qpid.proton.engine.BaseHandler.onReactorQuiesced(BaseHandler.java:87)
        at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:206)
        at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
        at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:309)
        at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:277)
        at com.microsoft.azure.servicebus.MessagingFactory$RunReactor.run(MessagingFactory.java:340)
        ... 1 more


---
In reply to: [104077500](https://github.com/Azure/azure-event-hubs-java/pull/66#discussion_r104077500) [](ancestors = 104077500)